deepreg/model/loss/image.py

```diff
@@ -3,214 +3,317 @@
 """
 import tensorflow as tf
 
+from deepreg.model.loss.util import NegativeLossMixin
+from deepreg.registry import REGISTRY
+
 EPS = tf.keras.backend.epsilon()
 
 
-def dissimilarity_fn(
-    y_true: tf.Tensor, y_pred: tf.Tensor, name: str, **kwargs
-) -> tf.Tensor:
+@REGISTRY.register_loss(name="ssd")
+class SumSquaredDifference(tf.keras.losses.Loss):
+    """
+    Sum of squared distance between y_true and y_pred.
+    y_true and y_pred have to be at least 1d tensor, including batch axis.
     """
-    Returns the calculated dissimilarity for each batch.
 
-    :param y_true: fixed_image, shape = (batch, f_dim1, f_dim2, f_dim3)
-    :param y_pred: warped_moving_image, shape = (batch, f_dim1, f_dim2, f_dim3)
-    :param name: name of the dissimilarity function
-    :param kwargs: absorb additional parameters
-    :return: shape = (batch,)
+    def __init__(
+        self, reduction=tf.keras.losses.Reduction.AUTO, name="SumSquaredDifference"
+    ):
+        """
+        :param reduction: using AUTO reduction, calling the loss like `loss(y_true, y_pred)` will return a scalar tensor.
+        :param name: name of the loss
+        """
+        super(SumSquaredDifference, self).__init__(reduction=reduction, name=name)
+
+    def call(self, y_true, y_pred):
+        """
+        :param y_true: shape = (batch, ...)
+        :param y_pred: shape = (batch, ...)
+        :return: shape = (batch,)
+        """
+        loss = tf.math.squared_difference(y_true, y_pred)
+        loss = tf.keras.layers.Flatten()(loss)
+        return tf.reduce_mean(loss, axis=1)
+
+
+class GlobalMutualInformation(tf.keras.losses.Loss):
     """
+    Differentiable global mutual information via Parzen windowing method.
+    y_true and y_pred have to be at least 4d tensor, including batch axis.
 
-    # shape = (batch, f_dim1, f_dim2, f_dim3, 1)
-    y_true = tf.expand_dims(y_true, axis=4)
-    y_pred = tf.expand_dims(y_pred, axis=4)
-    if name == "lncc":
-        return -local_normalized_cross_correlation(y_true, y_pred, **kwargs)
-    elif name == "ssd":
-        return ssd(y_true, y_pred)
-    elif name == "gmi":
-        return -global_mutual_information(y_true, y_pred)
-    else:
-        raise ValueError(
-            f"Unknown loss type {name}, supported losses are lncc, ssd and gmi."
-        )
+    Reference: https://dspace.mit.edu/handle/1721.1/123142, Section 3.1, equation 3.1-3.5, Algorithm 1
+    """
+
+    def __init__(
+        self,
+        num_bins: int = 23,
+        sigma_ratio: float = 0.5,
+        reduction=tf.keras.losses.Reduction.AUTO,
+        name="GlobalMutualInformation",
+    ):
+        """
+        :param num_bins: number of bins for intensity
```

Comment:
Can we add why num bins is default 23?

Reply:
This is probably beyond the topic of this PR, but it is an empirical default and reasonable to have a "balanced" number of bins - sufficiently distinct intensity values say for a 8bit image - and number of samples in each bin.

Reply:
I would still add this explanation to the docstring. Even if it is empirically determined, magic numbers are annoying to have and manage in code.

Reply:
I do not know... we can do a new issue to ask @acasamitjana

Reply:
Can you add this to the docstring as I have suggested? For the magic number managing reason, and raise a ticket besides this to discuss if this needs changning.

Reply:
done, https://github.com/DeepRegNet/DeepReg/issues/591